// molecule/codedom.js
export class Statement{

    constructor(){
        this.children = [];
    }

    indent(level){
        return '\t'.repeat(level);
    }

    toCode(indent){

    }

    childrenCode(indent){
        var result = '';
        for(let c of this.children){
            if(c != null) result += c.toCode(indent + 1) + '\n'
        }
        return result;
    }
}

export class Unit extends Statement{

    toCode(indent){
        return this.childrenCode(indent-1);
    }
    
}

export class Expr extends Statement{
    constructor(code){
        super()
        this.code = code;
    }
    toCode(indent){
        return this.code;
    }
}

Expr.toCode = function(expr, indent, stringify=true){
    return expr instanceof Expr ? expr.toCode(indent) : (stringify ? JSON.stringify(expr) : expr + '');
}

export class ExpressionStmt extends Statement{
    constructor(expr){
        super();
        this.expr = expr;
    }

    toCode(indent){
        return `${this.indent(indent)}${this.expr.toCode(indent)};`
    }
}

export class LiteralExpr extends Expr{     // const value
    constructor(value, type){
        super()
        this.value = value;
        this.type = type;
    }
    toCode(indent){
        switch(this.type){
        case 's': 
            return this.value == null ? 'null' : JSON.stringify(this.value);
        case 'n':            
        case 'b':
        case 'o':   // defered via JSON string
        case 'e':
            return this.value == null ? 'null' : this.value;
        case 'd':
            return this.value == null ? 'null' : `Date.parse(${JSON.stringify(this.value)})`;
        }
    }
    eval(){
        switch(this.type){
        case 's': 
            return this.value;
        case 'n':            
        case 'b':
        case 'o':   // defered via JSON string
        case 'e':
            return this.value == null ? null : JSON.parse(this.value);
        case 'd':
            return this.value == null ? null : Date.parse(this.value);
        }
    }
}

export class ObjectLiteralExpr extends Expr{
    constructor(define){
        super();
        this.define = define;
    }
    toCode(indent){
        var s = '{\n';
        for(let k in this.define){
            let expr = this.define[k];
            s += `${this.indent(indent + 1)}${k}: ${Expr.toCode(expr, indent + 3)},\n`
        }
        s += this.indent(indent) + '}';
        return s;
    }
}


export class AssignStmt extends Statement{
    constructor(left, right){
        super()
        this.left = left;
        this.right = right;     // right must be expr
    }
    
    toCode(indent){
        return this.indent(indent) + this.left + ' = ' + this.right.toCode(indent) + ';';
    }
}

export class ClassDecl extends Expr{
    constructor(name, extendsClassName, isExport){
        super()
        this.name = name;
        this.extendsClassName = extendsClassName;
        this.isExport = isExport;
    }

    toCode(indent){
        return `${this.indent(indent)}${this.isExport ? 'export ': ''}class ${this.name} extends ${Expr.toCode(this.extendsClassName, indent)} {\n${this.childrenCode(indent)}${this.indent(indent)}}`;
    }
}

export class FunctionDeclStmt extends ExpressionStmt{
    constructor(name, args){
        super(new FunctionDeclExpr(name, args))
    }
}

FunctionDeclStmt.fromStatements = function(name, statements){
    let r = new FunctionDeclStmt(name);
    r.children = statements;
    return r;
}

export class FunctionDeclExpr extends Expr{
    constructor(name, args, statements){
        super()
        this.name = name || '';
        this.args = args || '';
        this.children = statements || [];
    }

    toCode(indent){
        return `function ${this.name}(${this.args}) {\n${this.childrenCode(indent)}${this.indent(indent)}}`;
    }
}

FunctionDeclExpr.fromStatements = function(name, statements){
    let r = new FunctionDeclExpr(name);
    r.children = statements;
    return r;
}

export class MethodDecl extends Statement{
    constructor(name, args, statements){
        super()
        this.name = name || '';
        this.args = args || '';
        this.children = statements || [];
    }

    toCode(indent){
        return `${this.indent(indent)}${this.name}(${this.args || ''}) {\n${this.childrenCode(indent)}${this.indent(indent)}}`;
    }
}

export class ConstructorDecl extends Statement{
    constructor(){
        super()
    }

    toCode(indent){
        return `${this.indent(indent)}constructor() {\n${this.childrenCode(indent)}${this.indent(indent)}}`;
    }
}

export class LineStmt extends Statement{
    constructor(code){
        super()
        this.code = code;
    }

    toCode(indent){
        return this.indent(indent) + this.code + ';';
    }
}

export class PropAssignExprStmt extends Statement{
    constructor(elementName, propName, expr, isCustomProp){
        super()
        this.elementName = elementName
        this.propName = propName;
        this.expr = expr;
        this.isCustomProp = isCustomProp;
    }
    toCode(indent){
        if(this.isCustomProp){
            //return `${this.indent(indent)}${this.elementName}.setAttribute(${JSON.stringify(this.propName)}, ${this.expr.toCode()});`
            return `${this.indent(indent)}this.prop(${JSON.stringify(this.propName)}, ${this.expr.toCode(indent)});`
        } else {
            return `${this.indent(indent)}${this.elementName}.${this.propName} = ${this.expr.toCode(indent)};`
        }
    }
}

export class AttachEventExprStmt extends Statement{
    constructor(elementName, propName, expr, isCustomProp){
        super()
        this.elementName = elementName;   
        this.propName = propName.substr(2); // bypass 'on'
        this.expr = expr;
        this.isCustomProp = isCustomProp;
    }

    toCode(indent){
        if(! this.isCustomProp){            
            return `${this.indent(indent)}${this.elementName}.addEventListener(${JSON.stringify(this.propName)}, ${this.expr.toCode(indent)});`
        } else {        // TODO
            return `${this.indent(indent)}jQuery(${this.elementName}).on(${JSON.stringify(this.propName)}, ${this.expr.toCode(indent)});`
        }
    }
}

export class VarDeclStmt extends Statement{
    constructor(name, initExpr){
        super()
        this.name = name;
        this.initExpr = initExpr;        
    }

    toCode(indent){
        return `${this.indent(indent)}var ${this.name} = ${this.initExpr.toCode(indent)}`
    }
}

export class MethodInvokeExpr extends Expr{
    constructor(instance, methodName, args ){
        super()
        this.instance = instance;
        this.methodName = methodName;
        this.args = args;        
    }

    toCode(indent){
        let s = Expr.toCode(this.instance, indent, false);
        if(this.args instanceof ArrayLiteralExpr){
            return `${s}.${this.methodName}(${this.args.toCode(indent)})`
        } else {
            return `${s}.${this.methodName}(${this.args.map(a => Expr.toCode(a, indent)).join()})`;
        }
    }
}

export class MethodInvokeStmt extends ExpressionStmt{
    constructor(instance, methodName, args ){
        super(new MethodInvokeExpr(instance, methodName, args));
    }
}

export class FunctionInvokeExpr extends Expr{
    constructor(fun, args){
        super()
        this.fun = fun;
        this.args = args;        
    }

    toCode(indent){
        let s = Expr.toCode(this.fun, indent, false);
        if(this.args instanceof ArrayLiteralExpr){
            return `${s}(${this.args.toCode(indent)})`
        } else {
            return `${s}(${this.args.map(a => Expr.toCode(a, indent)).join()})`;
        }
    }
}

export class FunctionInvokeStmt extends ExpressionStmt{
    constructor(fun, args){
        super(new FunctionDeclExpr(fun, args))
    }
}

export class BracketExpr extends Expr{
    constructor(expr){
        super();
        this.expr = expr;
    }
    toCode(indent){
        return `(${this.expr.toCode(indent)})`
    }
}

export class LtEqExpr extends Expr{
    constructor(left, right){
        super();
        this.left = left;
        this.right = right;
    }
    toCode(indent){
        return Expr.toCode(this.left, indent) + '<=' + Expr.toCode(this.right, indent);
    }
}

export class ReturnStmt extends Statement{
    constructor(expr){
        super()
        this.expr = expr;
    }

    toCode(indent){
        return `${this.indent(indent)}return ${Expr.toCode(this.expr, indent)};`
    }
}

export class NewInstanceExpr extends Expr{
    constructor(className, constructArgs){
        super();
        this.className = className;
        this.constructArgs = constructArgs;
    }

    toCode(indent){
        return `new ${this.className}(${this.constructArgs.map(a => Expr.toCode(a, indent)).join()})`
    }
}

export class DefaultPropExpr extends NewInstanceExpr{
    // constructor(expression, type, isRuntime, isNative, echo)
    constructor(propName, type, isCustomProp, isExpr, isRuntime, isEcho, expr){
        super('MoleculeModule.Prop', [expr,  type, isRuntime ? true : false, (!isCustomProp) ? true : false, isEcho?true:false]);
    }
}

export class ArrayLiteralExpr extends Expr{
    constructor(){
        super();
        this.array = [];
    }

    push(element){
        this.array.push(element);
    }

    isEmpty(){
        return this.array.length == 0
    }

    extends(anotherArray){
        if(anotherArray instanceof ArrayLiteralExpr){
            Array.prototype.push.apply(this.array, anotherArray.array); 
        } else if(Array.isArray(anotherArray)){
            Array.prototype.push.apply(this.array, anotherArray);
        } else {
            throw new Error('type mismatch');
        }
    }

    toCode(indent){
        return `[${this.array.map(a => Expr.toCode(a, indent)).join()}]`;
    }
}

export class ExpandIteratorExpr extends Expr{
    constructor(expr){
        super();
        this.expr = expr;
    }

    toCode(indent){
        return '... ' + Expr.toCode(this.expr, indent, false);
    }
}

export class IfStmt extends Statement{

    constructor(branches){
        super();
        this.branches = branches;
    }

    toCode(indent){
        var s = ''
        var inIf = true;
        for (const branch of this.branches) {
            if(inIf){
                s += `${this.indent(indent)}if(${branch.cond.toCode(indent)}){\n`;
                inIf = false;
            } else if(branch.cond) {
                s += ` else if(${branch.cond.toCode(indent)}){\n`;
            }  else {
                s += ` else {\n`;
            }
            for(let stmt of branch.then){
                s += stmt.toCode(indent + 1) + '\n';
            }
            s += `${this.indent(indent)}}`
        }
        return s + '\n';
    }
}

export class ConstDeclStmt extends Statement{
    constructor(name, initExpr){
        super()
        this.name = name;
        this.initExpr = initExpr;        
    }

    toCode(indent){
        return `${this.indent(indent)}const ${this.name} = ${this.initExpr.toCode(indent)};`
    }
}

export class LambdaExpr extends Expr{
    constructor(args, statements){
        super()
        this.args = args || '';
        this.children = statements;
    }

    toCode(indent){
        return `(${this.args}) => {\n${this.childrenCode(indent)}${this.indent(indent)}}`;
    }
}

export class ForIteratorStmt extends Statement{
    
    constructor(iterator, container, body){
        super();
        this.iterator = iterator;
        this.container = container;
        this.children = body;
    }

    toCode(indent){
        const it = this.iterator.toCode(indent);

        return `${this.indent(indent)}for(${it.indexOf('.') == -1 ? 'const ' : ''}${this.iterator.toCode(indent)} of ${this.container.toCode(indent)}){\n${this.childrenCode(indent)}${this.indent(indent)}}`
    }
}

export class ForLoopStmt extends Statement{
    
    constructor(init, cond, step, body){
        super();
        this.init = init;
        this.cond = cond;
        this.step = step;
        this.children = body;
    }

    toCode(indent){
        return `${this.indent(indent)}for(${this.init ? this.init.toCode(indent) : ''};${this.cond ? this.cond.toCode(indent): ''}; ${this.step ? this.step.toCode(indent): ''}){\n${this.childrenCode(indent)}\n${this.indent(indent)}}\n`
    }
}

export class ConcatStringExpr extends Expr{

    constructor(...items){
        super()
        this.children = items;
    }

    extends(...items){
        return new ConcatStringExpr(...this.children.concat(items));
    }

    toCode(indent){
        let arr = [];
        var s = '';
        for(let c of this.children){
            if(c instanceof Expr){
                if(c instanceof LiteralExpr){
                    if(c.type == 's'){
                        s += c.value;
                    } else {
                        s += c.toCode(indent);
                    }
                } else {
                    if(s) arr.push(JSON.stringify(s));
                    s = '';
                    arr.push(c.toCode(indent));
                }
            } else {
                s += new String(c);
            }
        }
        if(s) arr.push(JSON.stringify(s));
        return arr.join('+');
        //return this.children.map(item => Expr.toCode(item, indent)).join('+');
    }
}

export class SwitchStmt extends Statement{

    constructor(cond, branches){
        super();
        this.cond = cond;
        this.branches = branches;
    }

    toCode(indent){
        var s = `${this.indent(indent)}switch(${Expr.toCode(this.cond, indent)}){\n`
        for (const branch of this.branches) {
            s += `${this.indent(indent)}case ${Expr.toCode(branch.cond, indent)}:\n`;
            for(let stmt of branch.then){
                s += stmt.toCode(indent + 1) + '\n';
            }
            s += `${this.indent(indent + 1)}break;\n`;
        }
        return s + `${this.indent(indent)}}`;
    }
}

export class ImportStmt extends Statement{
    constructor(items, module){
        super();
        this.items = items;
        this.module = module;
    }

    toCode(indent){
        if(this.items && this.items.length){
            return `${this.indent(indent + 1)}import ${this.items} from ${Expr.toCode(this.module, indent)};`  
        } else {
            return `${this.indent(indent + 1)}import ${Expr.toCode(this.module, indent)};`  
        }
    }
}

export class AwaitExpr extends Expr{
    constructor(expr){
        super();
        this.expr = expr;
    }

    toCode(indent){
        return `await ${this.expr.toCode(indent)}`;
    }
}

export class CreateElementExpr extends Expr{
    constructor(tag, attributes){
        super();
        this.tag = tag;
        this.attributes = attributes;
    }
    toCode(indent){
        return `jQuery(document.createElement('${this.tag}')).attr(${JSON.stringify(this.attributes)})`
    }
}

export class OrExpr extends Expr{
    constructor(expr1, expr2){
        super();
        this.expr1 = expr1;
        this.expr2 = expr2;
    }
    toCode(indent){
        return `${Expr.toCode(this.expr1, indent)} || ${Expr.toCode(this.expr2, indent)}`;
    }
}

export class Comment extends Expr{
    constructor(text, newLine){
        super();
        this.text = text;
        this.newLine = newLine;
    }
    toCode(indent){
        return `/*${this.text}*/${this.newLine? '\n': ''}`;
    }
}
